torchcde/interpolation_linear.py

```diff
@@ -82,7 +82,58 @@ def _linear_interpolation_coeffs_with_missing_values(t, x):
         return misc.cheap_stack(out_pieces, dim=0)
 
 
-def linear_interpolation_coeffs(x, t=None):
+def _prepare_rectilinear_interpolation(data, time_index):
+    """Prepares data for rectilinear interpolation.
+
+    This function performs the relevant filling and lagging of the data needed to convert raw data into a format such
+    standard linear interpolation will give the rectilinear interpolation.
+
+    Arguments:
+        x: tensor of values with first channel index being time, of shape (..., length, input_channels), where ... is
+            some number of batch dimensions.
+        time_index: integer giving the index of the time channel.
+
+    Example:
+        Suppose we have data:
+            data = [(t1, x1), (t2, NaN), (t3, x3), ...]
+        that we wish to interpolate using a rectilinear scheme. The key point is that this is equivalent to a linear
+        interpolation on
+            data_rect = [(t1, x1), (t2, x1), (t2, x1), (t3, x1), (t3, x3) ...]
+        This function simply performs the conversion from `data` to `data_rect` so that we can apply the inbuilt
+        torchcde linear interpolation scheme to achieve rectilinear interpolation.
+
+    Returns:
+        A tensor, now of shape (..., 2 * length - 1, input_channels] that can be fed to linear interpolation coeffs to
+            give rectilinear coeffs.
+    """
+    # Check time_index is of the correct format
+    n_channels = data.size(-1)
+    assert isinstance(time_index, int), "Index of the time channel must be an integer in [0, {}]".format(n_channels - 1)
+    assert 0 <= time_index < n_channels, "Time index must be in [0, {}], was given {}." \
+                                         "".format(n_channels - 1, time_index)
+    times = data[..., time_index]
+    # Check the specified channel dim does not violate time conditions
+    # The following asserts ensure that no nan value is sandwiched in between time values
+    assert not torch.isnan(times[..., 0]).any(), "One or more of the time indexes start with a nan value."
+    nan_then_time_value = torch.isnan((times[..., :-1] * times[..., 1:])[~torch.isnan(times[..., 1:])]).any()
+    assert not nan_then_time_value, "One of more of the time paths has a nan in between non-nan times which is " \
+                                    "not allowed."
+    # This then asserts times are increasing
+    time_diffs = times[..., 1:] - times[..., :1]
+    assert time_diffs[~torch.isnan(time_diffs)].min() > 0, "Found consecutive times with a difference <= 0." \
+                                                           "Please ensure that data[:, :, time_idx] corresponds to " \
+                                                           "times and ensure that they are increasing for each sample."
```

Comment:
This logic seems rather involved + moderately expensive to evaluate. I think for now let's just disallow nan times. We can revisit this if we need to later.